Issue #161: Document JPyInterpreter internals (#162)
optapy/optapy: jpyinterpreter-docs/src/modules/ROOT/pages/jpyinterpreter-introduction/jpyinterpreter-introduction.adoc

[[jpyinterpreterIntroduction]]
= JPyInterpreter Introduction
// Redirect to this page from .../docs/jpyinterpreter/latest.
:page-aliases: ../index.adoc
:doctype: book
:sectnums:
:icons: font

[[whatIsJpyinterpreter]]
== What is JPyInterpreter

> JPyInterpreter in a Python interpreter that runs in

Reviewer: ```suggestion
JPyInterpreter is a Python interpreter that runs in
```

the JVM that can be embedded in CPython to remove overhead
of calling Python code frequently from Java. This is mostly

> relevant for small functions, where the overhead decrease

Reviewer: ```suggestion
relevant for small functions, where the overhead decreases
```

performance by as much as 100 times.

It accomplishes this by creating Java class files from the
https://docs.python.org/3/glossary.html#term-bytecode[CPython bytecode], accessible https://docs.python.org/3/library/dis.html[via the dis module]. Both CPython Virtual Machine and the JVM are

> https://en.wikipedia.org/wiki/Stack_machine[stack machines], meaning each opcode read their parameters

Reviewer: ```suggestion
https://en.wikipedia.org/wiki/Stack_machine[stack machines], meaning each opcode reads its parameters
```

> and write their results to a stack. As such, JPyInterpreter can also be thought of as a translator of CPython bytecode to Java bytecode.

Reviewer: ```suggestion
and writes its results to a stack. As such, JPyInterpreter can also be thought of as a translator of CPython bytecode to Java bytecode.
```

[[jpyinterpreterStructure]]
== JPyInterpreter's Structure

JPyInterpreter's structure can be broken into several
components:

- The Python frontend, which is given Python function and classes and returns their corresponding Java versions.

>  The Python frontend has minimal code, and most responsible for passing the Python bytecode to Java and converting Python constants into Java constants.

Reviewer: ```suggestion
 The Python frontend has minimal code, and is mostly responsible for passing the Python bytecode to Java and converting Python constants into Java constants.
```

> - The Java opcode parser, which is given Python bytecode instructions and need to convert them to intermediary

Reviewer: ```suggestion
- The Java opcode parser, which is given Python bytecode instructions and needs to convert them to intermediary
```

 opcodes.

- A typeflow analyzer, which is given the Python function signature and the intermediary opcodes, and returns the stack metadata associated with each opcode. The stack metadata is useful for optimizing opcodes, such as `GET_ATTR` (which typically need to go through a complicated dict lookup process, but if it is a known field, it can be implemented by a single `getfield` java bytecode instruction)

> - Implementors, which writes the java bytecode instructions

Reviewer: ```suggestion
- Implementors, which write the java bytecode instructions
```
